Writing suggestions sometimes show up in the middle of previous paragraphs in `textarea`s

rr-codes opened the pull request:
#### 7685053aabfc692b2b35b2b6885eaa2af68da6ef
<pre>
Writing suggestions sometimes show up in the middle of previous paragraphs in `textarea`s
<a href="https://bugs.webkit.org/show_bug.cgi?id=273365">https://bugs.webkit.org/show_bug.cgi?id=273365</a>
<a href="https://rdar.apple.com/127129683">rdar://127129683</a>

Reviewed by Wenson Hsieh.

The logic that inserts the writing suggestions renderer was incorrectly assuming that the renderer
which corresponds to the node before the writing suggestion would always just be the first text
child node of the current renderer. While this is true for trivial `contenteditable` `div`s, which
have several `RenderBlock`s, and one `RenderText` per block, this is not true for `textarea`s,
which can and do have multiple `RenderText`s per block, nor is this true in the general case of
arbitrary HTML.

Since the node before the writing suggestions renderer is known already, its renderer is the correct
one to use, and doesn&apos;t make any assumptions about the number of children of a renderer.

* LayoutTests/editing/input/mac/writing-suggestions-textarea-multiple-lines-expected.txt: Added.
* LayoutTests/editing/input/mac/writing-suggestions-textarea-multiple-lines.html: Added.
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer):

Canonical link: <a href="https://commits.webkit.org/278140@main">https://commits.webkit.org/278140@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a5fc47d96322b99313aa8a1b42ad67261fed30d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49551 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28835 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52598 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52793 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/227 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51855 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34854 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26455 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40456 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51651 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26375 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42690 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21571 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23835 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43873 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7922 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45749 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44378 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54369 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24636 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20820 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47827 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25905 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46848 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10888 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26749 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25629 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on LayoutTests/editing/input/mac/writing-suggestions-textarea-multiple-lines.html:
```diff
@@ -0,0 +1,37 @@
+    testRunner.waitUntilDone();
+
+    await UIHelper.setInlinePrediction("name is", 2);
+    await UIHelper.ensurePresentationUpdate();
```
Test failure looks relevant. Is the output racy? Maybe `ensureStablePresentationUpdate`?

rr-codes replied:
It’s just because this feature isn’t available on older OSes

aprotyas replied:
oh, of course. Nvm

aprotyas commented on Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
```diff
@@ -304,19 +304,19 @@ void RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer(Rende
 
-    auto* firstChildText = dynamicDowncast<RenderText>(renderer.firstChild());
-    if (!firstChildText) {
+    auto* nodeBeforeWritingSuggestionsTextRenderer = dynamicDowncast<RenderText>(nodeBeforeWritingSuggestions->renderer());
```
This can be a `CheckedPtr`, right?
```suggestion
    CheckedPtr nodeBeforeWritingSuggestionsTextRenderer = dynamicDowncast<RenderText>(nodeBeforeWritingSuggestions->renderer());
```

aprotyas reviewed (commented):
This LGTM to me, but I'll delegate the r+ to someone more familiar with rendering code.

whsieh commented on LayoutTests/editing/input/mac/writing-suggestions-textarea-multiple-lines.html:
```diff
@@ -0,0 +1,37 @@
+
+    const shadowRoot = internals.shadowRoot(editor);
+
+    shadowRoot.childNodes[0].appendChild(document.createTextNode("My na"));
```
Is it necessary to reach into the shadow root, or can we type this out using `execCommand(…)`? If it works, I think that would more accurately reflect the user-facing steps in this bug.

rr-codes replied:
I initially tried using `execCommand`, but couldn't get the resulting shadow DOM to look quite like how it does in actuality, so I did this instead.

whsieh reviewed (approved):
Seems fine to me, but it would be good to have Alan or Antti double check.

rr-codes committed:
Writing suggestions sometimes show up in the middle of previous paragraphs in `textarea`s
https://bugs.webkit.org/show_bug.cgi?id=273365
rdar://127129683

Reviewed by Wenson Hsieh.

The logic that inserts the writing suggestions renderer was incorrectly assuming that the renderer
which corresponds to the node before the writing suggestion would always just be the first text
child node of the current renderer. While this is true for trivial `contenteditable` `div`s, which
have several `RenderBlock`s, and one `RenderText` per block, this is not true for `textarea`s,
which can and do have multiple `RenderText`s per block, nor is this true in the general case of
arbitrary HTML.

Since the node before the writing suggestions renderer is known already, its renderer is the correct
one to use, and doesn't make any assumptions about the number of children of a renderer.

* LayoutTests/editing/input/mac/writing-suggestions-textarea-multiple-lines-expected.txt: Added.
* LayoutTests/editing/input/mac/writing-suggestions-textarea-multiple-lines.html: Added.
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer):

Canonical link: https://commits.webkit.org/278140@main
